source/opt/propagator.cpp

```diff
@@ -0,0 +1,215 @@
+    // not to simulate anymore and add its SSA def-use edges for simulation.
+    DontSimulateAgain(instr);
+    if (instr->result_id() > 0) {
+      AddSSAEdges(instr->result_id());
+    }
+
+    /* If STMT transfers control out of its basic block, add
+       all outgoing edges to the work list.  */
+    if (instr->IsBlockTerminator()) {
+      ir::BasicBlock* block = ctx_->get_instr_block(instr);
+      for (const auto& e : bb_succs_.at(block)) {
+        AddControlEdge(e);
+      }
+    }
+    return;
+  } else if (status == Interesting) {
+    // If the instruction produced a new interesting value, add the SSA edge
+    // for its result ID.
+    if (instr->result_id() > 0) {
+      AddSSAEdges(instr->result_id());
+    }
+
+    // If there are multiple outgoing control flow edges and we know which one
+    // will be taken, add the destination block to the CFG work list.
+    if (dest_bb) {
+      blocks_.push(dest_bb);
+    }
+  }
+
+  // If there are no uses on the instruction whose definitions are simulated
+  // again, then this instruction will be never simulated again.
+  bool has_operands_to_simulate = false;
+  ir::BasicBlock* instr_bb = ctx_->get_instr_block(instr);
+  if (instr->opcode() == SpvOpPhi) {
+    for (uint32_t i = 2; i < instr->NumOperands(); i += 2) {
+      // Phi arguments come in pairs. Index 'i' contains the
+      // variable id, index 'i + 1' is the originating block id.
+      assert(i % 2 == 0 && i < instr->NumOperands() - 1 &&
+             "malformed Phi arguments");
+
+      uint32_t arg_id = instr->GetSingleWordOperand(i);
+      ir::Instruction* arg_def_instr = get_def_use_mgr()->GetDef(arg_id);
+      uint32_t in_label_id = instr->GetSingleWordOperand(i + 1);
+      ir::Instruction* in_label_instr = get_def_use_mgr()->GetDef(in_label_id);
+      ir::BasicBlock* in_bb = ctx_->get_instr_block(in_label_instr);
+      Edge edge(in_bb, instr_bb);
+
+      if (!IsEdgeExecutable(edge) || ShouldSimulateAgain(arg_def_instr)) {
+        has_operands_to_simulate = true;
+        break;
+      }
+    }
+  } else {
+    instr->ForEachInId([&has_operands_to_simulate, this](const uint32_t* use) {
+      ir::Instruction* def_instr = get_def_use_mgr()->GetDef(*use);
+      if (ShouldSimulateAgain(def_instr)) {
+        has_operands_to_simulate = true;
+        return;
+      }
+    });
+  }
+
+  if (!has_operands_to_simulate) {
+    DontSimulateAgain(instr);
+  }
+}
+
+template <typename ValueType>
+void SSAPropagator<ValueType>::Simulate(ir::BasicBlock* block) {
+  if (block == cfg_->pseudo_exit_block()) {
+    return;
+  }
+
+  // Always simulate Phi instructions, even if we have simulated this block
+  // before. We do this because Phi instructions receive their inputs from
+  // incoming edges. When those edges are marked executable, the corresponding
+  // operand can be evaluated.
+  block->ForEachPhiInst([this](ir::Instruction* instr) { Simulate(instr); });
```

Comment:
I don't understand why this is necessary.  Won't the phi nodes get added to "sss_edges_" when its operands are simulated.  So the phi nodes will get processed anyway if they need to.  Is that right?

Reply:
Phi nodes get operands enabled not only by SSA edges, a Phi operand becomes available also when the incoming CFG edge is marked executable.  That's why this code checks which operands have an executable predecessor.